### PR TITLE
Canvas: Add actions support

### LINKED ISSUE #90385: Canvas: Add actions functionality

### PR DESCRIPTION
**Pull request by adela-almasan**
https://github.com/user-attachments/assets/f480eb04-2512-4b3f-bbb2-dd0a64976101

attached is a test dashboard exercising Action interpolation in all the places (title, url, queryParam keys+values, header keys+values, and request body):

<details><summary>actions-test2.json</summary>

```json
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": {
          "type": "grafana",
          "uid": "-- Grafana --"
        },
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "fiscalYearStartMonth": 0,
  "graphTooltip": 0,
  "id": 338,
  "links": [],
  "panels": [
    {
      "datasource": {
        "name": "gdev-testdata",
        "type": "grafana-testdata-datasource",
        "uid": "PD8C576611E62080A"
      },
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds"
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 80
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 14,
        "w": 11,
        "x": 0,
        "y": 0
      },
      "id": 1,
      "options": {
        "infinitePan": false,
        "inlineEditing": false,
        "panZoom": false,
        "root": {
          "background": {
            "color": {
              "fixed": "transparent"
            }
          },
          "border": {
            "color": {
              "fixed": "dark-green"
            }
          },
          "constraint": {
            "horizontal": "left",
            "vertical": "top"
          },
          "elements": [
            {
              "actions": [
                {
                  "options": {
                    "body": "foo_var: ${foo_var} ---- A-series: ${__data.fields.A-series}",
                    "headers": [
                      [
                        "${foo_var}",
                        "${__data.fields.A-series}"
                      ],
                      [
                        "Content-Type",
                        "text/plain"
                      ]
                    ],
                    "method": "POST",
                    "queryParams": [
                      [
                        "${foo_var}",
                        "${__data.fields.A-series}"
                      ]
                    ],
                    "url": "/login/${foo_var}/${__data.fields.A-series}"
                  },
                  "title": "Poke  ${foo_var} with ${__data.fields.A-series}",
                  "type": "fetch"
                }
              ],
              "background": {
                "color": {
                  "fixed": "#D9D9D9"
                }
              },
              "border": {
                "color": {
                  "fixed": "dark-green"
                }
              },
              "config": {
                "align": "center",
                "color": {
                  "fixed": "#000000"
                },
                "size": 20,
                "text": {
                  "field": "",
                  "fixed": "",
                  "mode": "field"
                },
                "valign": "middle"
              },
              "constraint": {
                "horizontal": "left",
                "vertical": "top"
              },
              "links": [],
              "name": "Element 1",
              "oneClickMode": "off",
              "placement": {
                "height": 50,
                "left": 100,
                "rotation": 0,
                "top": 100,
                "width": 260
              },
              "type": "metric-value"
            }
          ],
          "name": "Element 1724970269352",
          "oneClickMode": "off",
          "placement": {
            "height": 100,
            "left": 0,
            "rotation": 0,
            "top": 0,
            "width": 100
          },
          "type": "frame"
        },
        "showAdvancedTypes": true
      },
      "pluginVersion": "11.3.0-pre",
      "targets": [
        {
          "datasource": {
            "type": "grafana-testdata-datasource",
            "uid": "PD8C576611E62080A"
          },
          "refId": "A",
          "scenarioId": "random_walk",
          "seriesCount": 1
        }
      ],
      "title": "Panel Title",
      "type": "canvas"
    }
  ],
  "preload": false,
  "schemaVersion": 39,
  "tags": [],
  "templating": {
    "list": [
      {
        "current": {
          "text": "foo123",
          "value": "foo123"
        },
        "name": "foo_var",
        "options": [
          {
            "selected": true,
            "text": "foo123",
            "value": "foo123"
          }
        ],
        "query": "foo123",
        "type": "custom"
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "browser",
  "title": "actions-test2",
  "uid": "bdwbw8mgvzqioa",
  "version": 8,
  "weekStart": ""
}

```
</details>

TODO
- [x] Variable suggestions
- [x] Variable interpolation
- [x] Add API call response feedback
- [x] Add feature toggle - `vizActions`
- [x] Update UX (after [datalinks UX PR](https://github.com/grafana/grafana/pull/91352) is merged)

Follow-ups
- [ ] Remove button API editor
- [ ] Refactor shared components between `Actions` and `DataLinks`
- [ ] Add "Test API" functionality

Fixes https://github.com/grafana/grafana/issues/90385

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/plugins/panel/canvas/editor/element/elementEditor.tsx [hunk 1]
```diff
@@ -120,21 +123,28 @@ export function getElementEditor(opts: CanvasEditorOptions): NestedPanelOptions<
...
-            { value: OneClickMode.Off, label: capitalize(OneClickMode.Off) },
-            { value: OneClickMode.Link, label: capitalize(OneClickMode.Link) },
-          ],
+          options: oneClickModeOptions,
         },
         defaultValue: OneClickMode.Off,
       });
 
       optionBuilder.addDataLinks(builder, ctx);
+      optionBuilder.addActions(builder, ctx);
```
Comment (nmarrs) [inline public/app/plugins/panel/canvas/editor/element/elementEditor.tsx:147]: MIght be cleaner / easier to understand in the future to just separate these changes into a single if block  - i.e.

```
const oneClickModeOptions = [...]
let oneClickCategory = "Data links"
let description = etc
if (actionsEnabled) {
  push action to oneClickModeOptions
  oneClickCategory += " and actions"
 description += " or action"
}

builder.addRadio({...})
  ```

**Review by nmarrs — COMMENTED** [review]
Let's gooooooo - I think this is really close, just a few of my classic nits / a few things we may want to address either now or in a follow-up soon

File: packages/grafana-ui/src/components/Actions/ActionButton.tsx [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { ActionModel, Field } from '@grafana/data';
+
+import { ButtonProps, Button } from '../Button';
+
+type ActionButtonProps = {
+  action: ActionModel<Field>;
+  buttonProps?: ButtonProps;
+};
```
Comment (baldm0mma) [inline packages/grafana-ui/src/components/Actions/ActionButton.tsx:8]: Another nit. 😂
```suggestion
};

```

File: .betterer.results [hunk 3]
```diff
@@ -196,6 +196,14 @@ exports[`better eslint`] = {
...
       [0, 0, 0, "Unexpected any. Specify a different type.", "1"]
     ],
+    "packages/grafana-data/src/types/action.ts:5381": [
+      [0, 0, 0, "Unexpected any. Specify a different type.", "0"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "1"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "2"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "3"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "4"],
+      [0, 0, 0, "Unexpected any. Specify a different type.", "5"]
+    ],
```
[the remark this answers is not in the document]
Comment (adela-almasan) [reply .betterer.results:206]: I had a conversation with @joshhunt about these yesterday. These files are not going to exist after refactoring, they are basically copies of data links files, and I'm planning to refactor in a future PR, as we get more clarity.

**Review by baldm0mma — APPROVED** [review]
🚀

File: packages/grafana-ui/src/components/Actions/ActionButton.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * @internal
+ */
+export function ActionButton({ action, buttonProps }: ActionButtonProps) {
```
[the remark this answers is not in the document]
Comment (adela-almasan) [reply packages/grafana-ui/src/components/Actions/ActionButton.tsx:12]: I think given that we're going to complete this in a couple of weeks with field level actions, doesn't make much sense to move the files around. 🤔

File: packages/grafana-data/src/types/action.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  // When a click callback exists, this is passed the raw mouse|react event
+  onClick?: (e: any, origin?: any) => void;
+}
+
+export interface ActionClickEvent<T = any> {
+  origin: T;
+  replaceVariables: InterpolateFunction | undefined;
+  e?: any; // mouse|react event
+}
```
[the remark this answers is not in the document]
Comment (adela-almasan) [reply packages/grafana-data/src/types/action.ts:32]: space police 😂

[the remark this answers is not in the document]
Comment (baldm0mma) [reply packages/grafana-data/src/types/action.ts:32]: I'm the worst, I know!!!

**Comment by adela-almasan** [comment]
@nmarrs I created some follow-up issues to address the [interpolation](https://github.com/grafana/grafana/issues/91588) and the [form validation](https://github.com/grafana/grafana/issues/91590). I don't think they're blockers.

**Review by nmarrs — APPROVED** [review]
🚢 🚀

File: packages/grafana-data/src/types/action.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+interface FetchOptions {
+  method: HttpRequestMethod;
+  url: string;
+  body?: string;
+  queryParams?: Array<[string, string]>;
+  headers?: Array<[string, string]>;
+}
+
+export enum ActionType {
+  Fetch = 'fetch',
```
Comment (Develer) [inline packages/grafana-data/src/types/action.ts:29]: maybe it's better to have just a constant instead of an enum in this case?

**Review by Develer — APPROVED** [review]
LGTM🤘🏼

File: packages/grafana-data/src/types/action.ts (hunk 6 above)
Comment (adela-almasan) [reply packages/grafana-data/src/types/action.ts:29]: added it as an enum as the list is going to grow soon 😁

File: packages/grafana-data/src/types/action.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  { label: HttpRequestMethod.POST, value: HttpRequestMethod.POST },
+  { label: HttpRequestMethod.PUT, value: HttpRequestMethod.PUT },
+  { label: HttpRequestMethod.GET, value: HttpRequestMethod.GET },
+];
+
+export const contentTypeOptions: SelectableValue[] = [
+  { label: 'application/json', value: 'application/json' },
+  { label: 'text/plain', value: 'text/plain' },
+  { label: 'application/javascript', value: 'application/javascript' },
+  { label: 'text/html', value: 'text/html' },
```
Comment (leeoniya) [inline packages/grafana-data/src/types/action.ts:48]: let's remove these. it's unlikely they are relevant in this "Actions" context.

File: packages/grafana-data/src/types/action.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+  { label: 'application/XML', value: 'application/XML' },
```
Comment (leeoniya) [inline packages/grafana-data/src/types/action.ts:49]:
```suggestion
  { label: 'application/xml', value: 'application/xml' },
```

File: packages/grafana-ui/src/components/Actions/SuggestionsInput.tsx [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+    }),
+    inputWrapper: css({
+      position: 'relative',
+    }),
+    // Wrapper with child selector needed.
+    // When classnames are applied to the same element as the wrapper, it causes the suggestions to stop working
+  };
+};
+
+export const SuggestionsInput = ({
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/Actions/SuggestionsInput.tsx:54]: this whole file looks like mostly copy/paste of

https://github.com/grafana/grafana/blob/main/public/app/features/transformers/suggestionsInput/SuggestionsInput.tsx

can you explain what changes were done, and maybe we can reuse the other one with some additions?

File: packages/grafana-data/src/types/dataLink.ts [hunk 10]
```diff
@@ -51,7 +51,6 @@ export interface DataLink<T extends DataQuery = any> {
   internal?: InternalDataLink<T>;
 
   origin?: DataLinkConfigOrigin;
-  sortIndex?: number;
```
Comment (leeoniya) [inline packages/grafana-data/src/types/dataLink.ts:54]: this is the "breaking" change. but it's not.

**Review by leeoniya — APPROVED** [review]

**Pull request merged**